File: server/app/Models/RentalApplication.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class RentalApplication extends Model
{
    use HasFactory;

    const STATUS_PENDING = 'pending';
    const STATUS_APPROVED = 'approved';
    const STATUS_REJECTED = 'rejected';
    const STATUS_MODIFIED_PENDING = 'modified-pending';
    const STATUS_MODIFIED_APPROVED = 'modified-approved';

    protected $fillable = [
        'user_id',
        'apartment_id',
        'check_in',
        'check_out',
        'message',
        'submission_attempt',
        'status',
        'rejected_reason',
        'submitted_at',
        'responded_at',
        'previous_status',
        'previous_data',
        'current_data',
        'modification_reason',
        'modification_submitted_at',
    ];

    protected $casts = [
        'check_in' => 'date',
        'check_out' => 'date',
        'submitted_at' => 'datetime',
        'responded_at' => 'datetime',
        'modification_submitted_at' => 'datetime',
        'previous_data' => 'array',
        'current_data' => 'array',
    ];

    public function user()
    {
        return $this->belongsTo(User::class, 'user_id');
    }

    public function apartment()
    {
        return $this->belongsTo(Apartment::class, 'apartment_id');
    }

    public function modifications()
    {
        return $this->hasMany(RentalApplicationModification::class, 'rental_application_id');
    }

    public function canBeModified(): bool
    {
        return in_array($this->status, [
            self::STATUS_PENDING,
            self::STATUS_APPROVED,
        ]);
    }

    public function getDiffWithPrevious(): array
    {
        if (!$this->previous_data || !$this->current_data) {
            return [];
        }

        $diff = [];
        foreach ($this->current_data as $key => $newValue) {
            $oldValue = $this->previous_data[$key] ?? null;
            if ($oldValue !== $newValue) {
                $diff[$key] = [
                    'old' => $oldValue,
                    'new' => $newValue,
                ];
            }
        }

        return $diff;
    }

    public function scopePending($query)
    {
        return $query->where('status', self::STATUS_PENDING);
    }

    public function scopeApproved($query)
    {
        return $query->where('status', self::STATUS_APPROVED);
    }

    public function scopeRejected($query)
    {
        return $query->where('status', self::STATUS_REJECTED);
    }

    public function scopeModifiedPending($query)
    {
        return $query->where('status', self::STATUS_MODIFIED_PENDING);
    }
}

File: client/lib/data/models/rental_application.dart
class RentalApplication {
  final String id;
  final String userId;
  final String apartmentId;
  final DateTime checkIn;
  final DateTime checkOut;
  final String? message;
  final int submissionAttempt;
  final String status;
  final String? rejectedReason;
  final DateTime submittedAt;
  final DateTime? respondedAt;
  final String? previousStatus;
  final Map<String, dynamic>? previousData;
  final Map<String, dynamic>? currentData;
  final String? modificationReason;
  final DateTime? modificationSubmittedAt;
  final Map<String, dynamic>? user;
  final Map<String, dynamic>? apartment;

  const RentalApplication({
    required this.id,
    required this.userId,
    required this.apartmentId,
    required this.checkIn,
    required this.checkOut,
    this.message,
    required this.submissionAttempt,
    required this.status,
    this.rejectedReason,
    required this.submittedAt,
    this.respondedAt,
    this.previousStatus,
    this.previousData,
    this.currentData,
    this.modificationReason,
    this.modificationSubmittedAt,
    this.user,
    this.apartment,
  });

  bool canBeModified() {
    return status == 'pending' || status == 'approved';
  }

  bool hasModification() {
    return status == 'modified-pending' || status == 'modified-approved';
  }

  factory RentalApplication.fromJson(Map<String, dynamic> json) {
    // Helper function to parse dates correctly without timezone issues
    DateTime parseDate(dynamic dateValue) {
      if (dateValue == null) throw Exception('Date value is null');
      final dateStr = dateValue.toString();
      
      // If it's just a date (YYYY-MM-DD), parse as local date without time
      if (dateStr.length == 10 && !dateStr.contains('T')) {
        final parts = dateStr.split('-');
        return DateTime(
          int.parse(parts[0]), // year
          int.parse(parts[1]), // month
          int.parse(parts[2]), // day
        );
      }
      
      // For datetime strings, parse normally
      return DateTime.parse(dateStr.replaceAll('Z', ''));
    }

    return RentalApplication(
      id: json['id'].toString(),
      userId: json['user_id'].toString(),
      apartmentId: json['apartment_id'].toString(),
      checkIn: parseDate(json['check_in']),
      checkOut: parseDate(json['check_out']),
      message: json['message'] as String?,
      submissionAttempt: json['submission_attempt'] ?? 0,
      status: json['status'] ?? 'pending',
      rejectedReason: json['rejected_reason'] as String?,
      submittedAt: parseDate(json['submitted_at']),
      respondedAt: json['responded_at'] != null 
          ? parseDate(json['responded_at']) 
          : null,
      previousStatus: json['previous_status'] as String?,
      previousData: json['previous_data'] as Map<String, dynamic>?,
      currentData: json['current_data'] as Map<String, dynamic>?,
      modificationReason: json['modification_reason'] as String?,
      modificationSubmittedAt: json['modification_submitted_at'] != null
          ? parseDate(json['modification_submitted_at'])
          : null,
      user: json['user'] as Map<String, dynamic>?,
      apartment: json['apartment'] as Map<String, dynamic>?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'apartment_id': apartmentId,
      'check_in': checkIn.toIso8601String(),
      'check_out': checkOut.toIso8601String(),
      'message': message,
      'submission_attempt': submissionAttempt,
      'status': status,
      'rejected_reason': rejectedReason,
      'submitted_at': submittedAt.toIso8601String(),
      'responded_at': respondedAt?.toIso8601String(),
      'previous_status': previousStatus,
      'previous_data': previousData,
      'current_data': currentData,
      'modification_reason': modificationReason,
      'modification_submitted_at': modificationSubmittedAt?.toIso8601String(),
      'user': user,
      'apartment': apartment,
    };
  }
}
